app/code/core/Mage/Core/Model/Config.php

```diff
@@ -501,13 +491,43 @@ public function getCache()
     }
 
     /**
-     * Get lock flag cache identifier
+     * Call before building and saving cache to ensure only one process can save the cache
      *
-     * @return string
+     * If failed to get cache lock:
+     *   - CLI: throws exception
+     *   - Other: 503 error
+     *
+     * @return void
+     * @throws Exception
      */
-    protected function _getCacheLockId()
+    public function getCacheSaveLock()
     {
-        return $this->getCacheId() . '.lock';
+        if (!Mage::app()->useCache('config')) {
+            return;
+        }
+        $connection = Mage::getSingleton('core/resource')->getConnection('core_write');
+        if (!$connection->fetchOne("SELECT GET_LOCK('core_config_cache_save_lock', ?)", [PHP_SAPI === 'cli' ? 60 : 3])) {
```

Comment:
```suggestion
        $lockWaitTime = getenv('MAGE_CONFIG_CACHE_LOCK_WAIT') ?: (PHP_SAPI === 'cli' ? 60 : 3);
        if (!$connection->fetchOne("SELECT GET_LOCK('core_config_cache_save_lock', ?)", [$lockWaitTime])) {
```

Reply:
This let's one override the default cache lock wait time with an environment variable, what do you think? Could also be moved to an XML config variable or just left hard-coded.
Regarding the default values I think sites with low traffic could afford to increase the wait time and sites with higher traffic need to keep it low. E.g. if you have 1000r/s then waiting 3 seconds will pile up 3000 requests whereas if you have 10r/s then only 30 requests is not a big deal.

Reply:
The `local.xml` file is already read in at this point but not yet merged with db config so it could go there, but then it wouldn't be dynamic.

The 3 seconds was picked trying to optimize between preventing the server(s) from getting so overloaded they start failing in other ways and how long it should reasonable take to save the cache. I think in most cases 3 seconds is enough time to save the cache so no requests would get dropped, but if the timeout were say 60, then if it did take over 3 seconds to save, that could allow a *lot* of requests to pile up. Also, in ecommerce a TTFB over 3 seconds is already atrocious so not sure how long it is worth waiting after that?

Ideally the system is configured to just drop connections after it becomes overloaded before they hit PHP anyway, but with AWS auto-scaling it could cause your container count to jump way up unnecessarily since this should just be a very temporary condition. 

Those are my thoughts. Happy to concede if you have another suggestion. 🙂

Reply:
Perhaps a good timeout would be the amount of time a normal config cache save *should* take on a healthy system - so if healthy then no requests will fail but if unhealthy they will start to abort to avoid a more catastrophic failure. Any idea what such a number would be? Maybe 6 seconds? If it takes more than that I would think either your systems are really slow or your config is holding way too much stuff and should be lightened up. :)